server/migrations/helpers.js

```diff
@@ -0,0 +1,50 @@
+function calculatePreferences(user) {
+  const updatedPreferences = {
+    config: '',
+    sort: 0,
+    star: user?.preferences?.star || [],
+    complete: user?.preferences?.complete || [],
+  }
+  // Update complete
+  if (
+    !!user.preferences &&
+    user.preferences.complete &&
+    Object.keys(user.preferences.complete).length &&
+    !Array.isArray(user.preferences.complete)
+  ) {
+    updatedPreferences.complete = Object.keys(user.preferences.complete).map(
+      (siteName) => {
+        return {
+          site: siteName,
+          completedSubjects: user.preferences.complete[siteName],
+        }
+      }
+    )
+  }
+
+  //Update Starred
+  if (
+    !!user.preferences &&
+    user.preferences.star &&
+    Object.keys(user.preferences.star).length &&
+    !Array.isArray(user.preferences.star)
+  ) {
+    updatedPreferences.star = Object.keys(user.preferences.star).map(
+      (siteName) => {
+        return {
+          site: siteName,
+          starredSubjects: user.preferences.star[siteName],
+        }
+      }
+    )
+  }
```

Comment:
I think we can consolidate some of this logic to make it easier to read

```suggestion
  const isUserCompletePreferencesMigrated = Array.isArray(user.preferences?.complete)
  const isUserStarredPreferencesMigrated = Array.isArray(user.preferences?.star)
  const isUserMigrated = isUserCompletePreferencesMigrated && isUserStarredPreferencesMigrated
  
  if (isUserMigrated) {
    return updatedPreferences
  }
 
  const userCompletedSites = Object.keys(user.preferences?.complete || [])
  const userStarredSites = Object.keys(user.preferences?.star || [])

  updatedPreferences.complete = userCompletedSites.map((siteName) => ({
    site: siteName,
    completedSubjects: user.preferences.complete[siteName] || [],
  }))

  updatedPreferences.star = userStarredSites.map((siteName) => ({
    site: siteName,
    completedSubjects: user.preferences.star[siteName] || [],
  }))
```

Reply:
https://github.com/AMP-SCZ/dpdash/pull/241